include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -1060,58 +1082,365 @@ namespace Differentiation
+      NumberTypes::none,
+      typename std::enable_if<
+        std::is_arithmetic<ScalarType>::value ||
+        (boost::is_complex<ScalarType>::value &&
+         std::is_arithmetic<typename internal::RemoveComplexWrapper<
+           ScalarType>::type>::value)>::type>
     {
+      /**
+       * The internal number type code
+       */
+      static constexpr enum NumberTypes type_code = NumberTypes::none;
+
+      // The clang compiler does not seem to like these
+      // variables being defined as constant expressions
+      // (the tests <adolc|sacado>/ad_number_traits_02 will
+      // fail with linking errors). However, GCC complains
+      // about the use of non-constant expressions in
+      // std::conditional.
+#  ifdef __clang__
+
+      /**
+       * A flag to indicate whether the number is of
+       * the taped variety or not
+       */
+      static const bool is_taped;
+
+
+      /**
+       * A flag to indicate whether the number is of
+       * the tapeless variety or not
+       */
+      static const bool is_tapeless;
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a real value
+       */
+      static const bool is_real_valued;
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a complex value
+       */
+      static const bool is_complex_valued;
+
+
+      /**
+       * The number of directional derivatives that can be
+       * taken with this (floating point) number
+       */
+      static const unsigned int n_supported_derivative_levels;
+
+#  else
+
+      /**
+       * A flag to indicate whether the number is of
+       * the taped variety or not
+       */
+      static constexpr bool is_taped = false;
+
+
+      /**
+       * A flag to indicate whether the number is of
+       * the tapeless variety or not
+       */
+      static constexpr bool is_tapeless = false;
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a real value
+       */
+      static constexpr bool is_real_valued =
+        (!boost::is_complex<ScalarType>::value);
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a complex value
+       */
+      static constexpr bool is_complex_valued =
+        !(NumberTraits<ScalarType, NumberTypes::none>::is_real_valued);
+
+
+      /**
+       * The number of directional derivatives that can be
+       * taken with this (floating point) number
+       */
+      static constexpr unsigned int n_supported_derivative_levels = 0;
+
+#  endif
+
+
       /**
        * Underlying floating point value type.
        * This could real-valued or complex-valued.
        */
       using scalar_type = ScalarType;
 
-      static ScalarType
-      get_directional_derivative(const ScalarType & /*x*/,
-                                 const unsigned int /*direction*/)
+
+      /**
+       * Type for real numbers
```

Comment:
Here and below, maybe end the sentence with a period? I think we're reasonably consistent in this, even if the documentation consists of only a half-sentence as here.